### PR TITLE
Day pass

### PR DESCRIPTION
**Pull request by Mark-Powers**
Enables day passes. 

User work flow steps:
1. A user assigns their trovi artifact to a project, and enables reproducibility.
2. This creates a new project for reproducibility with a small allocation request. Right now this is manually approved, but I left code in commented out that will automatically approve an initial allocation.
3. Another user can click "request day pass" under launch on the artifact. This takes them to a form.
4. All managers on a project receive mail with the submitted request, and can review it.
5. If approved, an invite is sent to the user who requested the day pass. When they use the invite, their time starts.
6. After their time ends, they are automatically removed from the project. If they are the 10th user removed from this project, the allocation expiration date is set to now, and an email is sent to the PI requesting more information about their day pass usage.

Deployed to https://dev.chameleoncloud.org/

### CODE REVIEW

**Comment by Mark-Powers** [comment]
Right now day passes are generated by the PI manually adding new users with a duration. It would be good to have a conversation to refine this idea.

File: projects/models.py
Comment (Mark-Powers) [reply projects/models.py:122]: I have preferred BEYOND_DURATION, as this signifies that an invite period is over, rather than the invite expiration date was exceeded.

**Comment by Mark-Powers** [comment]
Deployed to https://dev.chameleoncloud.org/

File: projects/tasks.py
Comment (zhenz) [inline projects/tasks.py:74]: maybe set this number as a variable in Django setting?

File: projects/views.py
Comment (zhenz) [inline projects/views.py:90]: I pushed a patch last night. could you rebase? and I think we can get rid of this check.

**Review by zhenz — APPROVED** [review]
A lot of changes! LGTM! I left a couple of comments. I released a patch for the PI delegate changes. Could you please rebase? And I might overwrite your deployment on dev last night. sorry!

File: sharing_portal/urls.py
Comment (diurnalist) [inline sharing_portal/urls.py:14]: Nit: in general our paths follow REST conventions, so this would just be `path("requests", ...)`

File: sharing_portal/templates/sharing_portal/share.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/share.html:74]:
```suggestion
            and reproduce your artifact. If you enable reproducibility requests,
```

File: sharing_portal/templates/sharing_portal/share.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/share.html:76]: I think maybe the information most relevant to the PI is:

"if granted, they will be given a small temporary allocation, which will be tied to your project's allocation. While usage by reproducing users will not count towards your project's main allocation, the PI is nonetheless accountable for their responsible usage of Chameleon resources. All reproducibility requests therefore require PI approval."

Or maybe something simpler.

File: sharing_portal/templates/sharing_portal/review_day_pass.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/review_day_pass.html:3]: Don't need this if `render_bundle` is not used in the template.

File: sharing_portal/templates/sharing_portal/request_day_pass.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/request_day_pass.html:39]: maybe instead of "that will convince them to approve your request" say "relevant to your request", to make it sound less like this is their opportunity to crack open their copy of _How to Win Friends and Influence People_ :D

File: sharing_portal/templates/sharing_portal/request_day_pass.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/request_day_pass.html:3]: Don't need render_bundle

File: sharing_portal/templates/sharing_portal/list_day_pass_requests.html
Comment (diurnalist) [inline sharing_portal/templates/sharing_portal/list_day_pass_requests.html:40]: Is this needed here? I think this was only put on some other templates because JS on the page needed access to the model data.

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:356]: Cool, I wanted to make sure we still show the button for folks not logged in

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:392]: 🚨  Make sure the sharing key is included in the launch URL as well

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:481]: Ah, I see you're using this trick elsewhere :+1:

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:493]: Something to tweak later... it may make sense to make the mail pretty innocuous and put this information in the review page instead. Otherwise we are maintaining a lot of text in various places.

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:649]: denied

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:883]: hmm?

File: sharing_portal/views.py
Comment (diurnalist) [inline sharing_portal/views.py:895]: I was thinking we should get rid of this "source" thing at some point, but this is actually the first time it may be useful :D -- we could say "Day pass" or something, idk. Probably not worth it.

**Review by diurnalist — CHANGES_REQUESTED** [review]
Huzzah! Great work here. You are a real Django form slinger now. I left several comments, but highlighted the ones that I feel should be addressed before merge with a 🚨 . The main thing I am wondering about is the additional model field - these things are expensive to change later, so we should get it right now.

Watch out for the empty CSS/build files - `projects/static/projects/css/tmpwsnqxp2e` should be deleted before merging.

File: projects/forms.py
Comment (Mark-Powers) [reply projects/forms.py:212]: I'm opting to completely remove this from the form. I think it would be confusing to users to see this when adding members.

File: sharing_portal/views.py
Comment (Mark-Powers) [reply sharing_portal/views.py:895]: I've changed it to day pass, it is a simple change that probably doesn't hurt anything.

File: sharing_portal/views.py
Comment (Mark-Powers) [reply sharing_portal/views.py:493]: I've removed this in favor of text on the review page.

File: sharing_portal/views.py
Comment (Mark-Powers) [reply sharing_portal/views.py:392]: Fixed

File: sharing_portal/views.py
Comment (Mark-Powers) [reply sharing_portal/views.py:356]: Yep!

File: sharing_portal/templates/sharing_portal/share.html
Comment (Mark-Powers) [reply sharing_portal/templates/sharing_portal/share.html:76]: Updated.

File: sharing_portal/migrations/0016_day_pass_fields.py
Comment (Mark-Powers) [reply sharing_portal/migrations/0016_day_pass_fields.py:45]: Yes, I figured this was best to prevent spam and also to give more of an identity to a user requesting one.

File: sharing_portal/views.py
Comment (super-cooper) [inline sharing_portal/views.py:284]: Should we catch the exception that this can throw?

File: projects/views.py
Comment (Mark-Powers) [reply projects/views.py:159]: Fixed.

File: projects/views.py
Comment (Mark-Powers) [reply projects/views.py:150]: Yes, this should be the same as duration is not null, and I can use that instead.

File: projects/views.py
Comment (Mark-Powers) [reply projects/views.py:123]: Yes. It also prevents there from being 2 active day passes in case a user requests 2 and is approved for both.

File: projects/views.py
Comment (Mark-Powers) [reply projects/views.py:178]: Fixed.

File: projects/migrations/0011_update_invitations.py
Comment (Mark-Powers) [reply projects/migrations/0011_update_invitations.py:37]: I've removed these, I think they were from the old way I was doing durations.

**Review by super-cooper — CHANGES_REQUESTED** [review]
Looks great! Some of my comments might be outdated at this point.

Also, I think you accidentally added the file `projects/static/projects/css/tmpwsnqxp2e`

File: projects/forms.py
Comment (super-cooper) [reply projects/forms.py:208]: Oh it's a string my b

File: sharing_portal/views.py
Comment (Mark-Powers) [reply sharing_portal/views.py:893]: This is copied from how we do it when creating a project.

File: projects/tasks.py
Comment (super-cooper) [inline projects/tasks.py:70]:
```suggestion
                if approved_requests >= settings.DAY_PASS_LIMIT:
```

Logically, this change is unnecessary, but I think it's better to have this still be handled if some bug in the future somehow allows people to approve users past this limit. With the current logic, if you managed to somehow get over 10 approved requests, you would be allowed to approve all future requests ad-infinitum.

**Comment by Mark-Powers** [comment]
Deployed to dev.chameleoncloud.org

File: projects/tasks.py
Comment (super-cooper) [inline projects/tasks.py:74]: Does this indicate that something is wrong, or that something unexpected is going on? Since we get the invitations from `get_invitations_beyond_duration`, this should mean it's a valid invitation, each of which should have one associated day pass request, right? Should it somehow be possible for an invitation to not have an associated day pass request?

If this state is not expected, I think it would be good to log something here.

File: projects/views.py
Comment (super-cooper) [inline projects/views.py:129]: This may be the first time I've ever seen a finally block that was actually necessary.

**Review by super-cooper — APPROVED** [review]
LGTM! A couple small changes I'd like, but I don't want to hold you up on getting another approval from me after you make them.

File: projects/tasks.py
Comment (Mark-Powers) [reply projects/tasks.py:70]: Since the allocation is expired at 10, they will have to re-request one which makes sure we are aware of their usage. This lets us gauge the popularity of day pass for now. If we did >=, they would need to request an allocation for every new user after this initial allocation.

File: projects/models.py
Comment (diurnalist) [inline projects/models.py:178]: 🙌🏻

File: projects/views.py
Comment (diurnalist) [reply projects/views.py:170]: I agree the comment is helpful, and I think it's one that is worth having.

File: sharing_portal/static/sharing_portal/day_pass.css
Comment (diurnalist) [inline sharing_portal/static/sharing_portal/day_pass.css:1]: 🙇🏻

File: sharing_portal/views.py
Comment (diurnalist) [reply sharing_portal/views.py:404]: Thanks!

File: sharing_portal/views.py
Comment (diurnalist) [reply sharing_portal/views.py:893]: Ah right, this is a weird legacy TAS (TACC Accounting System) thing, I think.

**Review by diurnalist — APPROVED** [review]
Great work, and congrats! :shipit: 🥳

**Pull request merged**
